pkg/action/hooks.go

```diff
@@ -107,6 +124,17 @@ func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent,
 	return nil
 }
 
+// getHookLog gets the log from the pod associated with the given hook, which is expected to be a test hook.
+func getHookLog(client kubernetes.Interface, rel *release.Release, hook *release.Hook) (release.HookLog, error) {
+	req := client.CoreV1().Pods(rel.Namespace).GetLogs(hook.Name, &v1.PodLogOptions{})
+	responseBody, err := req.DoRaw(context.Background())
```

Comment:
Why the change from `req.Stream` to `req.DoRaw`?

Reply:
In the current implementation, [the logs are acquired](https://github.com/helm/helm/blob/03d9e7ca62e428b9263dcf19542691b0531c0ccc/pkg/action/release_testing.go#L114) and then immediately [written to an `io.Writer`](https://github.com/helm/helm/blob/03d9e7ca62e428b9263dcf19542691b0531c0ccc/pkg/action/release_testing.go#L120). With our suggested changes, the acquisition and the writing are decoupled such that no `io.Writer` is in scope [when the logs are acquired](https://github.com/helm/helm/blob/d478b411b953f69e2f5d8a5505f49ba2dedc8524/pkg/action/hooks.go#L129). Instead, [`getHookLog` _returns_ the hook log](https://github.com/helm/helm/blob/d478b411b953f69e2f5d8a5505f49ba2dedc8524/pkg/action/hooks.go#L127), which can then be [written to an `io.Writer`](https://github.com/helm/helm/blob/d478b411b953f69e2f5d8a5505f49ba2dedc8524/pkg/action/release_testing.go#L115) or handled in any other suitable way.

Note that, unlike the current implementation, `getHookLog`

  * can _acquire_ logs without the side effect of _writing_ to an `io.Writer`, and
  * is used at multiple locations ([in `hooks.go`](https://github.com/helm/helm/blob/d478b411b953f69e2f5d8a5505f49ba2dedc8524/pkg/action/hooks.go#L95) and [in `release_testing.go`](https://github.com/helm/helm/blob/d478b411b953f69e2f5d8a5505f49ba2dedc8524/pkg/action/release_testing.go#L110)).

It might be possible to throw out `req.DoRaw` and shoehorn `req.Stream` into its place, but I don't know how, and it seems to me that doing so would just make `getHookLog` unnecessarily complex.

That being said, I can't assess e.g. performance implications of using `req.DoRaw` over `req.Stream` – I usually think in types and values rather than side effects and low-level implementation details.

Reply:
https://pkg.go.dev/k8s.io/client-go@v0.22.4/rest#Request.Stream

https://pkg.go.dev/k8s.io/client-go@v0.22.4/rest#Request.DoRaw

Seems like `req.Stream` performs some higher-level operations like handling non-2xx response codes whereas `req.DoRaw` does nothing of the sort. As a result, I'm very concerned that this change will not capture the same error cases Kubernetes might throw at us as before.

I'd highly suggest looking into the differences between these two API calls before swapping them out.

Reply:
Yes. I'll see if I can revert to `req.Stream`.

Reply:
Done.